src/com/amazon/ion/system/IonReaderBuilder.java

```diff
@@ -157,6 +168,110 @@ private IonCatalog validateCatalog()
         return catalog != null ? catalog : new SimpleCatalog();
     }
 
+    /**
+     * <p>
+     * Determines whether the IonReader will allow incremental reading of binary Ion data. When enabled, if
+     * {@link IonReader#next()} returns {@code null} at the top-level, it indicates that there is not enough data
+     * in the stream to complete a top-level value. The user may wait for more data to become available in the stream
+     * and call {@link IonReader#next()} again to continue reading. Unlike the non-incremental reader, the incremental
+     * reader will never throw an exception due to unexpected EOF during {@code next()}. If, however,
+     * {@link IonReader#close()} is called when an incomplete value is buffered, an {@link IonException} will be raised.
+     * </p>
+     * <p>
+     * There is no incremental text IonReader, so for text data a non-incremental IonReader will always be returned
+     * regardless of the value of this option.
```

Comment:
This seems like a potential stumbling block, but I'm not sure I have a satisfying solution. One mitigation that comes to mind is allowing users to query their `IonReader` to see if it conforms to the parameters from the builder; i.e. add a `boolean isIncremental()` method to the interface.

Users that don't care about the behavioral differences between the incremental and non-incremental readers (those who are in it for the performance improvements) would never call it. Users that absolutely require incremental reading (who are rare) could query their reader as soon as it's constructed.

I notice that the existing `IonReader` interface doesn't allow users to check whether it's reading text or binary, so maybe there's some philosophical reason we don't provide that transparency.

Reply:
> I think the biggest challenge with adding to the IonReader interface is that it would make this a backward-incompatible change.

I forgot we're on such an old JVM version that default interface method implementations aren't supported. 😞 

`isIncremental()` was my preferred path. I agree that throwing an exception complicates testing. We could also spell out the limitation in the builder method? `setIncrementalForBinaryReaders(true)`? I don't like it, but none of the options feel great.

Reply:
I'm pretty ambivalent about changing the builder method name vs leaving it as-is and keeping the text vs. binary difference explicit in the JavaDoc, which I think is pretty clear.

The pro of renaming the method is that it makes it easier for users who refuse to read the JavaDocs for the methods they use to avoid a rare pitfall. But they'd probably need to read the JavaDoc to understand what the feature even does...

The con is that making a binary-specific method pollutes the builder API if we ever do add an incremental text reading option. We'd probably end up having a deprecated binary-only method and a supported unified method. Then we have to add annoying logic to handle the case where someone specifies both and they conflict...

I guess I'm leaning slightly toward keeping the name and documentation as-is, but if you really think changing the name is better I can do that.

Reply:
That's a good point. Here's why having future incremental text reading support enabled by `withIncrementalReadingEnabled` doesn't feel like a breaking change to me. Let's say the user enables the option right now and processes text data that contains an incomplete top-level value. Currently, the text reader will throw an exception when trying to step past the incomplete value. If we add incremental text reading, the reader will now return null in that case. There are two ways the user may be handling this:
1. Since incremental reading was not previously supported, the user is still treating null at the top-level as the end of stream indicator, so their code will break its loop and eventually close the reader. At that point, the unexpected EOF exception will be thrown because the reader was closed with a value partially buffered. So the point at which the exception is raised changes, but it is still raised. Or,
2. The user is treating the reader as an incremental reader because they may also be processing binary data incrementally. They get unexpected EOFs for text when they'd get nulls for binary, and therefore they must have logic set up for both. In this case, text just starts following the same code paths as binary, which would be desired.

This isn't a one-way door, either. If we ship this and later add incremental text support and decide we can't reuse `withIncrementalReadingEnabled`, we can create `withIncrementalReadingEnabledForText` and `withIncrementalReadingEnabledForBinary` (where `withIncrementalReadingEnabledForBinary` just does what `withIncrementalReadingEnabled` does today), and deprecate `withIncrementalReadingEnabled`.

In the meantime, I'll update the JavaDoc on `withIncrementalReadingEnabled` to explicitly call out that incremental reading of text is not currently supported, but may be enabled by this option in the future.

Reply:
I agree with this assessment. Thanks for going through it!